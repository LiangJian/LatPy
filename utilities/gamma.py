import numpy as np

g1 = np.matrix([[   0,   0,   0, -1j],
                [   0,   0, -1j,   0], 
                [   0, +1j,   0,   0], 
                [ +1j,   0,   0,   0]])

g2 = np.matrix([[   0,   0,   0,  -1],
                [   0,   0,  +1,   0], 
                [   0,  +1,   0,   0], 
                [  -1,   0,   0,   0]])

g3 = np.matrix([[   0,   0, -1j,   0],
                [   0,   0,   0, +1j], 
                [ +1j,   0,   0,   0], 
                [   0, -1j,   0,   0]])

g4 = np.matrix([[  +1,   0,   0,   0],
                [   0,  +1,   0,   0], 
                [   0,   0,  -1,   0], 
                [   0,   0,   0,  -1]])

g5 = np.matrix([[   0,   0,  +1,   0],
                [   0,   0,   0,  +1], 
                [  +1,   0,   0,   0], 
                [   0,  +1,   0,   0]])

g0 = np.matrix([[  +1,   0,   0,   0],
                [   0,  +1,   0,   0], 
                [   0,   0,  +1,   0], 
                [   0,   0,   0,  +1]])

gi = [g0, g1, g2, g3, g4, g5]
gip = []
giv = []
for ig in gi:
    for i in range(4):
        for j in range(4):
            if ig[i, j] != 0.0:
                gip.append(j)
                giv.append(ig[i, j])

gip = np.array(gip)
giv = np.array(giv)
gip = gip.reshape(6, 4)
giv = giv.reshape(6, 4)

g15 = g1*g5
g25 = g2*g5
g35 = g3*g5
g45 = g4*g5

g51 = -g15
g52 = -g25
g53 = -g35
g54 = -g45

g5i = [g51, g52, g53, g54]
g5ip = []
g5iv = []
for ig in g5i:
    for i in range(4):
        for j in range(4):
            if ig[i, j] != 0.0:
                g5ip.append(j)
                g5iv.append(ig[i, j])

g5ip = np.array(g5ip)
g5iv = np.array(g5iv)
g5ip = g5ip.reshape(4, 4)
g5iv = g5iv.reshape(4, 4)

gi5 = [g15, g25, g35, g45]
gi5p = []
gi5v = []
for ig in gi5:
    for i in range(4):
        for j in range(4):
            if ig[i, j] != 0.0:
                gi5p.append(j)
                gi5v.append(ig[i, j])

gi5p = np.array(gi5p)
gi5v = np.array(gi5v)
gi5p = gi5p.reshape(4, 4)
gi5v = gi5v.reshape(4, 4)

g = gi + gi5
gp = np.concatenate([gip,gi5p])
gv = np.concatenate([giv,gi5v])

