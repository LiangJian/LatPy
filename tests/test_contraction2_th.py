import torch as th
import numpy as np
import time
from gamma import *
from opt_einsum import contract


def read_peram(peram_dir, conf_id, Nt, Nev, Nev_use, t_source):
	peram = []
	for d_source in range(4):
		peram.append(np.fromfile("%s/perams.%s.%i.%i"%(peram_dir,conf_id,d_source,t_source), dtype='c16'))
	peram = np.array(peram)	
	peram=peram.reshape(4,Nt,Nev,4,Nev)  # d_source, t_sink, ev_source, d_sink, ev_sink
	peram=peram.transpose(1,4,2,3,0)  # t_sink, ev_sink, ev_souce, d_sink, d_source
	peram=peram[:,:Nev_use,:Nev_use,:,:]
	return peram


def read_eig(eig_dir, conf_id, Nt, Nx, Nev, Nev_use, t_source):
	eigv=np.fromfile("%s/eigvecs_t%03d_%d"%(eig_dir,t_source,conf_id), dtype='c16')
	eigv=eigv.reshape(Nev,Nx**3,3)  # ev, nx**3, color
	eigv=eigv[:Nev_use,...]
	return eigv


def read_vvv(vvv_dir, conf_id, Nev, Nev_use, t_source):
	vvv=np.fromfile("%s/VVV.t%02d.conf%d"%(vvv_dir,t_source,conf_id), dtype='c16')
	vvv=vvv.reshape(Nev,Nev,Nev)
	vvv=vvv[:Nev_use,:Nev_use,:Nev_use]
	return vvv


peram_dir = '/beegfs/group/lqcd/perambulators/beta6.41_mu-0.2295_ms-0.2050_L32x96/light/test_Nev600/7400/'
eig_dir = '/beegfs/home/liuming/LapH/data/eigensystem/conf_32x96_b6.41_2295_2050/hyp_smear/Nev600/7400/'
vvv_dir='/beegfs/home/liuming/LapH/data/VVV/conf_32x96_b6.41_2295_2050/7400/'
conf_id = 7400
Nev = 600
Nev_use = 200
Nx = 32
Nt = 96

cg5 = g2 @ g4 @ g5
# cg5 = th.tensor(cg5, device='cuda')
cg5 = th.tensor(cg5.astype('c8'), device='cuda')
Gamma = (g0 + g4)/2.
# Gamma = th.tensor(Gamma, device='cuda')
Gamma = th.tensor(Gamma.astype('c8'), device='cuda')

# vvv = []
# for t_source in range(0,Nt,1):
# 	eigv = read_eig(eig_dir, conf_id, Nt, Nx, Nev, Nev_use, t_source)  # i_ev, spatial, color
# 	eigv = th.tensor(eigv, device='cuda')
# 	temp = contract(eigv,('i_ev', 'x', 'i_c1'), eigv,('j_ev', 'x', 'i_c2'), eigv,('k_ev', 'x', 'i_c3'), ('i_ev', 'j_ev', 'k_ev', 'i_c1', 'i_c2', 'i_c3'))
# 	vvv.append(contract(temp, ('i_ev', 'j_ev', 'k_ev', 'i_c1', 'i_c2', 'i_c3'), epsilon, ('i_c1', 'i_c2', 'i_c3'), ('i_ev', 'j_ev', 'k_ev')))
# 	print(t_source,end='\t', flush=True)
# vvv = np.array(vvv)
# vvv = vvv.reshape(Nt,Nev_use,Nev_use,Nev_use)
# np.save('vvv_%d'%conf_id,vvv)
# vvv = np.load(('vvv_%d.npy'%conf_id))
# print("vvv done")  # t_source, ev_1, ev_2, ev_3 
# print(vvv.shape)
# vvv= th.tensor(vvv, device='cuda')

vvv = []
st = time.time()
for it in range(0, 96, 1):
	vvv.append(read_vvv(vvv_dir, conf_id, 300, Nev_use, it))
ed = time.time()
print('read vvv used %.4f'%(ed - st), 's.')
vvv = np.array(vvv)
print(vvv.shape)
st = time.time()
# vvv = th.tensor(vvv, device='cuda')
vvv = th.tensor(vvv.astype('c8'), device='cuda')  # to single
ed = time.time()
print('copy vvv to gpu used %.4f'%(ed - st), 's.')
print('GPU memory used %.4f'%(th.cuda.memory_allocated()/1024/1024/2014),'GB.')

res = []

dt_source = 4
dt_sink = 4

for t_source in range(0,Nt,dt_source):
	st = time.time()
	peram0=read_peram(peram_dir, conf_id, Nt, Nev, Nev_use, t_source)
	print(peram0.shape) # t_sink, ev_sink, ev_souce, d_sink, d_source
	ed = time.time()
	print('read peram with t0 = %d done, used %.4f'%(t_source, ed - st), 's.')

	st = time.time()
	for t_sink in range(0,Nt,dt_sink):
		peram = peram0[t_sink:t_sink+dt_sink,...]
		# P = th.tensor(peram, device='cuda')
		P = th.tensor(peram.astype('c8'), device='cuda')  # to single

		Qu = contract(P, ('t_sink', 'ev_sink', 'ev_souce', 'd_sink', 'd_source'), cg5, ('d_source', 'd2_source'),
				('t_sink', 'ev_sink', 'ev_souce', 'd_sink', 'd2_source'))  # cg5 onto source
		
		Qd = contract(P, ('t_sink', 'ev_sink', 'ev_souce', 'd_sink', 'd_source'), cg5, ('d2_sink', 'd_sink'),
				('t_sink', 'ev_sink', 'ev_souce', 'd2_sink', 'd_source'))  # cg5 onto sink
	
		V = contract(vvv[t_source, ...].conj(), ('j_ev', 'l_ev', 'n_ev'), Qu, ('t_sink', 'i_ev', 'j_ev', 'alpha_1', 'beta_2'),
				('t_sink', 'i_ev', 'l_ev', 'n_ev', 'alpha_1', 'beta_2'))

		T = contract(V, ('t_sink', 'i_ev', 'l_ev', 'n_ev', 'alpha_1', 'beta_2'), Qd, ('t_sink', 'k_ev', 'l_ev', 'alpha_1', 'beta_2'),
				('t_sink', 'i_ev', 'k_ev', 'n_ev'))

		V = contract(vvv[t_sink:t_sink+dt_sink,...], ('t_sink', 'i_ev', 'k_ev', 'm_ev'), T, ('t_sink', 'i_ev', 'k_ev', 'n_ev'), 
				('t_sink', 'm_ev', 'n_ev'))

		T = contract(Gamma, ('gamma_2', 'gamma_1'), P, ('t_sink', 'm_ev', 'n_ev', 'gamma_1', 'gamma_2'), 
				('t_sink', 'm_ev', 'n_ev'))

		# P1 = contract(V, ('t_sink', 'm_ev', 'n_ev'), T, ('t_sink', 'm_ev', 'n_ev'), ('t_sink'))  # there is a bug
		P1 = contract('ijk,ijk->i', V, T)
		res.append(np.array(P1.cpu()))

		#------------
		
		V = contract(vvv[t_source, ...].conj(), ('j_ev', 'l_ev', 'n_ev'), P, ('t_sink', 'i_ev', 'j_ev', 'alpha_1', 'gamma_2'),
				('t_sink', 'i_ev', 'l_ev', 'n_ev', 'alpha_1', 'gamma_2'))

		T = contract(V, ('t_sink', 'i_ev', 'l_ev', 'n_ev', 'alpha_1', 'gamma_2'), Qd, ('t_sink', 'k_ev', 'l_ev', 'alpha_1', 'beta_2'),
				('t_sink', 'i_ev', 'k_ev', 'n_ev', 'gamma_2', 'beta_2'))

		V = contract(T, ('t_sink', 'i_ev', 'k_ev', 'n_ev', 'gamma_2', 'beta_2'), Qu, ('t_sink', 'm_ev', 'n_ev', 'gamma_1', 'beta_2'),
				('t_sink', 'i_ev', 'k_ev', 'm_ev', 'gamma_2', 'gamma_1'))

		T = contract(Gamma, ('gamma_2', 'gamma_1'), V, ('t_sink', 'i_ev', 'k_ev', 'm_ev', 'gamma_2', 'gamma_1'), 
				('t_sink', 'i_ev', 'k_ev', 'm_ev'))

		# P2 = contract(vvv[t_sink:t_sink+dt_sink,...], ('t_sink', 'i_ev', 'k_ev', 'm_ev'), T, ('t_sink', 'i_ev', 'k_ev', 'm_ev'), ('t_sink'))
		P2 = contract('ijkl,ijkl->i', vvv[t_sink:t_sink+dt_sink,...], T)
		res.append(np.array(P2.cpu()))

	ed = time.time()
	print('contraction for peram with t0 = %d done, used %.4f'%(t_source, ed - st), 's.')
	print('max GPU memory used %.4f'%(th.cuda.max_memory_allocated()/1024/1024/2014),'GB.')

res = np.array(res)
res = res.reshape(Nt//dt_source,Nt//dt_sink,2,dt_sink)
res = res.transpose(2,0,1,3)
res = res.reshape(2,Nt//dt_source,Nt)

# for i in range(Nt):
# 	print(i, '%.8e'%res[0, 0, i].real, '%.8e'%res[0, 0, i].imag, '%.8e'%res[1, 0, i].real, '%.8e'%res[1, 0, i].imag)

res = res[0,...] + res[1,...]
np.save('res.%d'%conf_id, res)

for i in range(Nt):
	print(i, '%.8e'%res[0, i].real)

