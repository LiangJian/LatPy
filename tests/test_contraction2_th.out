read vvv used 128.8476 s.
(96, 200, 200, 200)
copy vvv to gpu used 4.0154 s.

GPU memory used 2.9096 GB.

(96, 200, 200, 4, 4)
read peram with t0 = 0 done, used 29.3666 s.
contraction for peram with t0 = 0 done, used 19.0671 s.
max GPU memory used 10.7075 GB.

(96, 200, 200, 4, 4)
read peram with t0 = 4 done, used 30.0085 s.
contraction for peram with t0 = 4 done, used 18.6578 s.
max GPU memory used 10.7075 GB.

(96, 200, 200, 4, 4)
read peram with t0 = 8 done, used 28.4535 s.
contraction for peram with t0 = 8 done, used 18.6670 s.
max GPU memory used 10.7075 GB.

(96, 200, 200, 4, 4)
read peram with t0 = 12 done, used 29.0232 s.
contraction for peram with t0 = 12 done, used 18.6855 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 16 done, used 29.1016 s.
contraction for peram with t0 = 16 done, used 18.6900 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 20 done, used 28.1892 s.
contraction for peram with t0 = 20 done, used 18.6962 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 24 done, used 30.0972 s.
contraction for peram with t0 = 24 done, used 18.6930 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 28 done, used 30.3569 s.
contraction for peram with t0 = 28 done, used 18.6933 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 32 done, used 28.7986 s.
contraction for peram with t0 = 32 done, used 18.7178 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 36 done, used 28.3685 s.
contraction for peram with t0 = 36 done, used 18.7034 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 40 done, used 29.7289 s.
contraction for peram with t0 = 40 done, used 18.7022 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 44 done, used 28.3824 s.
contraction for peram with t0 = 44 done, used 18.7325 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 48 done, used 28.0687 s.
contraction for peram with t0 = 48 done, used 18.7277 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 52 done, used 29.0566 s.
contraction for peram with t0 = 52 done, used 18.7282 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 56 done, used 29.2288 s.
contraction for peram with t0 = 56 done, used 18.7291 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 60 done, used 28.9067 s.
contraction for peram with t0 = 60 done, used 18.7224 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 64 done, used 29.5423 s.
contraction for peram with t0 = 64 done, used 18.7217 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 68 done, used 29.0151 s.
contraction for peram with t0 = 68 done, used 18.7195 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 72 done, used 28.1966 s.
contraction for peram with t0 = 72 done, used 18.7313 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 76 done, used 28.7431 s.
contraction for peram with t0 = 76 done, used 18.7133 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 80 done, used 28.7370 s.
contraction for peram with t0 = 80 done, used 18.7039 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 84 done, used 29.5161 s.
contraction for peram with t0 = 84 done, used 18.7156 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 88 done, used 27.4226 s.
contraction for peram with t0 = 88 done, used 18.7252 s.
max GPU memory used 10.7075 GB.
(96, 200, 200, 4, 4)
read peram with t0 = 92 done, used 28.5092 s.
contraction for peram with t0 = 92 done, used 18.7140 s.
max GPU memory used 10.7075 GB.
0 7.33117223e-01
1 1.94394469e-01
2 8.11906606e-02
3 3.87825184e-02
4 1.98226664e-02
5 1.06065013e-02
6 6.21781358e-03
7 3.70537490e-03
8 2.18043686e-03
9 1.26131147e-03
10 8.31937068e-04
11 5.21057053e-04
12 3.25100729e-04
13 1.88853999e-04
14 1.03107624e-04
15 6.41189836e-05
16 3.10703617e-05
17 2.15331202e-05
18 8.60604814e-06
19 -4.35094307e-06
20 -1.15099801e-05
21 -1.12008247e-05
22 -9.40884092e-06
23 -5.83000974e-06
24 -3.50905634e-06
25 -2.58208047e-06
26 -1.60845002e-06
27 -7.91021307e-07
28 3.13859857e-07
29 8.16267118e-07
30 5.20302194e-07
31 3.69437849e-08
32 2.67549410e-07
33 3.15464177e-07
34 1.79243216e-07
35 -3.02224734e-08
36 -1.19112244e-07
37 -5.88291869e-08
38 -4.64559236e-09
39 7.31169436e-09
40 -1.47509773e-08
41 -1.64516845e-09
42 8.85987284e-09
43 1.89576994e-08
44 6.16291089e-08
45 5.51624701e-08
46 5.69695864e-08
47 1.83685351e-08
48 -2.04281356e-08
49 -3.22356861e-08
50 -8.08000351e-08
51 -5.45017365e-08
52 3.15493338e-08
53 7.63299539e-08
54 9.04682693e-08
55 1.57860399e-07
56 1.42834210e-07
57 8.19526207e-08
58 -1.18590773e-08
59 -6.72864502e-08
60 -1.43998989e-07
61 -2.28076914e-07
62 -3.82151768e-07
63 -1.42963472e-07
64 -1.95079423e-07
65 -2.07449659e-07
66 -2.44010295e-07
67 -3.19479085e-07
68 1.13713554e-07
69 5.32957756e-07
70 8.28302461e-07
71 1.30425974e-06
72 1.82655333e-06
73 1.34348761e-06
74 2.08776487e-06
75 -8.47836930e-07
76 -2.17707270e-06
77 6.95343658e-07
78 -5.59528587e-07
79 4.23243910e-06
80 8.04561205e-06
81 1.19155993e-05
82 5.33244292e-06
83 -1.63778714e-06
84 1.08621007e-05
85 3.05469548e-05
86 6.42668238e-05
87 1.24792670e-04
88 2.89557822e-04
89 6.20724924e-04
90 1.27168815e-03
91 2.72446452e-03
92 6.26430567e-03
93 1.43413153e-02
94 3.69924307e-02
95 1.07881010e-01
